### PR TITLE
Changes to AzureResourceGroupDeployment to fix #3868

### LINKED ISSUE #3868: AzureResourceGroupDeployment does not support deallocating VMs
[AzureResourceGroupDeployment](https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/AzureResourceGroupDeployment) does not support deallocating a VM. 

It would be very nice to add a check box to the stop virtual machine action to deallocate the VM. The box should default to false so it will remain backwards compatible when scripts update.

### PR DESCRIPTION
**Pull request by leftler**
* Bumped version number to 2.1.0 because this adds functionality in a backwards compatible manner.
 * Added a `deallocate` function to azure-arm-compute.ts
 * Added a checkbox to "Deallocate VM" that is only shown when action is set to `Stop virtual machines`
 * Added a `deallocateVMOnStop` boolean to `AzureRGTaskParameters`
 * Added branching code to check if `deallocateVMOnStop` was true or not inside VirtualMachine.ts
 * Added notes to README.md
 * Renamed `Stopped VMs` test case to `Stopped VMs without deallocating`, added a value for `deallocateVMOnStop = false`, and fixed a typo in the assert.
 * Added new test case 'Stopped VMs with deallocating'
 * Added mocks to VirtualMachines.js

This defaults the new Deallocate VM checkbox to false so it will be backwards compatible with 2.0.5

### CODE REVIEW

File: Tasks/AzureResourceGroupDeployment/task.loc.json [hunk 1]
```diff
@@ -77,6 +77,14 @@
       }
     },
     {
+      "name": "deallocateVMOnStop",
+      "type": "boolean",
+      "label": "ms-resource:loc.input.label.deallocateVMOnStop",
+      "groupName": "AzureDetails",
+      "visibleRule": "action = Stop virtual machines",
+      "defaultValue": false
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/task.loc.json]: Build your task, so that the changes will reflect here

File: Tasks/AzureResourceGroupDeployment/Tests/mock_node_modules/azure-arm-compute/package.json [hunk 2]
```diff
@@ -84,6 +84,10 @@
     {
       "name": "Zavery, Amar",
       "email": "amzavery@microsoft.com"
+    },
+    {
+      "name": "Chamberlain, Scott",
+      "email": "leftler@gmail.com"
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/Tests/mock_node_modules/azure-arm-compute/package.json]: This is not necessary

Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/Tests/mock_node_modules/azure-arm-compute/package.json]: This is mock library, which is used for tests alone

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 3]
```diff
@@ -77,6 +77,14 @@
             }
         },
         {
+          "name": "deallocateVMOnStop",
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/task.json]: Remove this and add this an action

File: Tasks/AzureResourceGroupDeployment/Tests/L0.ts [hunk 4]
```diff
@@ -492,6 +493,24 @@ describe('Azure Resource Group Deployment', function () {
             done(error);
         }
     });
+    it('Stopped VMs with deallocating', (done) => {
+        let tp = path.join(__dirname, 'VMOperations.js');
+        process.env["operation"] = "Stop";
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/Tests/L0.ts]: You will have to change this to "StopWithDeallocate"

File: Tasks/AzureResourceGroupDeployment/README.md [hunk 5]
```diff
@@ -52,6 +52,10 @@ The parameters of the task are described in details, including examples, to show
  
  * **Location**\*: The location where the resource group will be created. If an existing resource group is updated, then this parameter will be ignored.
 
+The following parameter is shown when the selected action is to stop virtual machines:
+
+ * **Deallocate VM**\*: Once the virtual machine is stopped it will be deallocated.
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/README.md]: You'll have to move this to actions

File: Tasks/AzureResourceGroupDeployment/operations/VirtualMachine.ts [hunk 6]
```diff
@@ -39,8 +39,16 @@ export class VirtualMachine {
                             client.virtualMachines.start(this.taskParameters.resourceGroupName, vmName, callback(vmName));
                             break;
                         case "Stop":
-                            console.log(tl.loc("VM_Stop", vmName));
-                            client.virtualMachines.powerOff(this.taskParameters.resourceGroupName, vmName, callback(vmName));
+                            if (this.taskParameters.deallocateVMOnStop)
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/operations/VirtualMachine.ts]: You'll have to fix here as well

File: Tasks/AzureResourceGroupDeployment/task.json (hunk 3 above)
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/task.json]: "Stop and deallocate virtual machines" should be the label in actions dropdown

**Comment by leftler** [comment]
@ds-ms Please let me know if there are any other changes you need done.

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 7]
```diff
@@ -13,11 +13,11 @@
...
     "version": {
         "Major": 2,
-        "Minor": 0,
-        "Patch": 6
+        "Minor": 1,
+        "Patch": 0
     },
     "demands": [],
-    "minimumAgentVersion": "2.0.0",
+    "minimumAgentVersion": "2.1.0",
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/task.json]: Why did you change this version?

Comment (leftler) [inline Tasks/AzureResourceGroupDeployment/task.json]: Because I updated the task version to 2.1.0 [here](https://github.com/Microsoft/vsts-tasks/pull/3869/files#diff-0bf716797e060f6c813008991524e17fL15) to follow [Semantic Versioning](http://semver.org/) because my change does "add functionality in a backwards-compatible manner".

Maybe I do not understand what minimumAgentVersion means, I thought it meant the minimum version number of the Task that the agent was allowed to use, the agent must have at least version 2.1 so that it includes the new logic that was added to azure-arm-compute.ts. If my understanding is incorrect can you please point me to the proper documentation for that field?

Comment (leftler) [inline Tasks/AzureResourceGroupDeployment/task.json]: Ok, I found some documentation and I now understand. I have reverted the change.

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 8]
```diff
@@ -374,4 +374,4 @@
         "CouldNotDetectEncoding": "Could not detect encoding of file '%s'",
         "ErrorWhileParsingParameter": "There was an error while overriding '%s' parameter because of '%s'."
     }
-}
\ No newline at end of file
+}
```
Comment (leftler) [inline Tasks/AzureResourceGroupDeployment/task.json]: The web editor on GitHub added the extra line at the end. Do you need me to remove it?

Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/task.json]: That's alright

**Comment by ds-ms** [comment]
Did you test if this works?

**Comment by leftler** [comment]
Yes, it works fine. Here is a copy of the debug logs of it running on my personal VSTS account against my personal Azure subscription
[logs_61.zip](https://github.com/Microsoft/vsts-tasks/files/909082/logs_61.zip)

**Comment by ds-ms** [comment]
@leftler Thanks for your contribution 👍
